google-cloud-logging/src/main/java/com/google/cloud/logging/LoggingHandler.java

```diff
@@ -457,28 +469,64 @@ void write(LogEntry entry, WriteOption... options) {
           reportError(null, ex, ErrorManager.FLUSH_FAILURE);
         }
         break;
+
       case ASYNC:
       default:
-        ApiFutures.addCallback(getLogging().writeAsync(entryList, options), new ApiFutureCallback<Void>() {
-          @Override
-          public void onSuccess(Void v) {}
-
-          @Override
-          public void onFailure(Throwable t) {
-            if (t instanceof Exception) {
-              reportError(null, (Exception) t, ErrorManager.FLUSH_FAILURE);
-            } else {
-              reportError(null, new Exception(t), ErrorManager.FLUSH_FAILURE);
-            }
-          }
-        });
+        final ApiFuture<Void> writeFuture = getLogging().writeAsync(entryList, options);
+        writeLock.lock();
+        try {
+          pendingWrites.add(writeFuture);
+        } finally {
+          writeLock.unlock();
+        }
+        ApiFutures.addCallback(
+            writeFuture,
+            new ApiFutureCallback<Void>() {
+              @Override
+              public void onSuccess(Void v) {
+                writeLock.lock();
+                try {
+                  pendingWrites.remove(writeFuture);
+                } finally {
+                  writeLock.unlock();
+                }
+              }
+
+              @Override
+              public void onFailure(Throwable t) {
+                try {
+                  if (t instanceof Exception) {
+                    reportError(null, (Exception) t, ErrorManager.FLUSH_FAILURE);
+                  } else {
+                    reportError(null, new Exception(t), ErrorManager.FLUSH_FAILURE);
+                  }
+                } finally {
+                  onSuccess(null);
```

Comment:
Calling `onSuccess` from `onFailure` could be a hazard for future maintenance - it looks correct now, but anything added to `onSuccess` later would affect `onFailure` too, which could be very surprising. I would prefer both methods to call a shared separate function, or just duplicate the code.